File: packages/signals/src/text.ts
/**
 * Text helpers. Line-for-line twin: apps/api/signalyze_api/engine/text.py.
 * All lengths are counted in Unicode code points so both runtimes agree.
 */
import languages from '../data/languages.json';
import stopwords from '../data/stopwords.json';

/**
 * Languages with a phrase dictionary and a tone lexicon, in the order of
 * data/languages.json (which is also the tie order of the stopword vote).
 */
export const ENGINE_LANGUAGES = [
  'en',
  'es',
  'pt',
  'fr',
  'de',
  'it',
  'tr',
  'nl',
  'pl',
  'id',
  'vi',
  'sv',
  'ru',
  'uk',
  'ar',
  'ja',
  'zh',
  'ko',
] as const;
export type EngineLanguage = (typeof ENGINE_LANGUAGES)[number];
/** Detection result: a dictionary language, or 'other' when no dictionary applies. */
export type DetectedLanguage = EngineLanguage | 'other';
export type Script = 'latin' | 'cyrillic' | 'arabic' | 'japanese' | 'han' | 'hangul';
/** 'word': space tokens and word-boundary phrases; 'substring': plain substrings (no word separators). */
export type MatchMode = 'word' | 'substring';

export interface LanguageEntry {
  code: EngineLanguage;
  script: Script;
  matchMode: MatchMode;
}

const SCRIPTS: readonly Script[] = ['latin', 'cyrillic', 'arabic', 'japanese', 'han', 'hangul'];
const MATCH_MODES: readonly MatchMode[] = ['word', 'substring'];

function isOneOf<T extends string>(values: readonly T[], value: string): value is T {
  return (values as readonly string[]).includes(value);
}

/** The language table of data/languages.json, validated once at load. */
export const LANGUAGE_TABLE: readonly LanguageEntry[] = languages.languages.map((entry) => {
  const { code, script, matchMode } = entry;
  if (
    !isOneOf(ENGINE_LANGUAGES, code) ||
    !isOneOf(SCRIPTS, script) ||
    !isOneOf(MATCH_MODES, matchMode)
  ) {
    throw new Error(`languages.json: unknown entry ${JSON.stringify(entry)}`);
  }
  return { code, script, matchMode };
});

/** Builds a table with one value per engine language. */
export function perLanguage<T>(build: (lang: EngineLanguage) => T): Record<EngineLanguage, T> {
  const out: Partial<Record<EngineLanguage, T>> = {};
  for (const lang of ENGINE_LANGUAGES) out[lang] = build(lang);
  return out as Record<EngineLanguage, T>;
}

/** How phrases and lexicon entries are matched per language (from languages.json). */
export const MATCH_MODE: Record<EngineLanguage, MatchMode> = perLanguage((lang) => {
  const entry = LANGUAGE_TABLE.find((e) => e.code === lang);
  if (entry === undefined) throw new Error(`languages.json has no entry for ${lang}`);
  return entry.matchMode;
});

function languagesOfScript(script: Script): readonly EngineLanguage[] {
  const codes = LANGUAGE_TABLE.filter((e) => e.script === script).map((e) => e.code);
  if (codes.length === 0) throw new Error(`languages.json has no ${script} language`);
  return codes;
}

/** Candidates of the stopword vote per script, in tie order; the first is the default. */
const LATIN_LANGUAGES = languagesOfScript('latin');
const CYRILLIC_LANGUAGES = languagesOfScript('cyrillic');

const STOPWORDS: Record<EngineLanguage, readonly string[]> = stopwords;

/** token -> languages whose stopword list contains it (ENGINE_LANGUAGES order), for an O(tokens) vote. */
const STOPWORD_LANGUAGES: ReadonlyMap<string, readonly EngineLanguage[]> = (() => {
  const map = new Map<string, EngineLanguage[]>();
  for (const lang of ENGINE_LANGUAGES) {
    for (const word of STOPWORDS[lang]) {
      const list = map.get(word);
      if (list === undefined) map.set(word, [lang]);
      else list.push(lang);
    }
  }
  return map;
})();

/**
 * NFKC, lower-case, keep letters/digits/whitespace, collapse whitespace, trim.
 * Python twin: unicodedata.normalize('NFKC', s).lower(); re.sub(r'[^\w\s]|_', ' '); re.sub(r'\s+', ' ').strip()
 */
export function normalizeText(text: string): string {
  return text
    .normalize('NFKC')
    .toLowerCase()
    .replace(/[^\p{L}\p{N}\s]+/gu, ' ')
    .replace(/\s+/gu, ' ')
    .trim();
}

export function codePointLength(text: string): number {
  let n = 0;
  for (const _ of text) n += 1;
  return n;
}

export function tokenize(normalized: string): string[] {
  if (normalized === '') return [];
  return normalized.split(' ');
}

type ScriptBucket = 'kana' | 'hangul' | 'han' | 'arabic' | 'cyrillic' | 'latin' | 'other';

/**
 * Script bucket of one code point of normalised text, or null for characters
 * that carry no script (space, ASCII digits). The ranges are Unicode blocks;
 * the Python twin lists exactly the same ranges.
 */
function scriptOf(cp: number): ScriptBucket | null {
  if (cp <= 0x7f) return cp >= 0x61 && cp <= 0x7a ? 'latin' : null;
  if (
    (cp >= 0xc0 && cp <= 0x24f) ||
    (cp >= 0x1e00 && cp <= 0x1eff) ||
    (cp >= 0x2c60 && cp <= 0x2c7f) ||
    (cp >= 0xa720 && cp <= 0xa7ff)
  ) {
    return 'latin';
  }
  if (cp >= 0x400 && cp <= 0x52f) return 'cyrillic';
  if (
    (cp >= 0x600 && cp <= 0x6ff) ||
    (cp >= 0x750 && cp <= 0x77f) ||
    (cp >= 0x8a0 && cp <= 0x8ff) ||
    (cp >= 0xfb50 && cp <= 0xfdff) ||
    (cp >= 0xfe70 && cp <= 0xfefe)
  ) {
    return 'arabic';
  }
  if (
    (cp >= 0x3040 && cp <= 0x30ff) ||
    (cp >= 0x31f0 && cp <= 0x31ff) ||
    (cp >= 0xff66 && cp <= 0xff9f)
  ) {
    return 'kana';
  }
  if (
    (cp >= 0x1100 && cp <= 0x11ff) ||
    (cp >= 0x3130 && cp <= 0x318f) ||
    (cp >= 0xa960 && cp <= 0xa97f) ||
    (cp >= 0xac00 && cp <= 0xd7ff)
  ) {
    return 'hangul';
  }
  if (
    (cp >= 0x3005 && cp <= 0x3007) ||
    (cp >= 0x3400 && cp <= 0x4dbf) ||
    (cp >= 0x4e00 && cp <= 0x9fff) ||
    (cp >= 0xf900 && cp <= 0xfaff) ||
    (cp >= 0x20000 && cp <= 0x3134f)
  ) {
    return 'han';
  }
  return 'other';
}

/** Most stopword hits among `candidates`; ties resolve to the earlier candidate, no hits to the first. */
function stopwordVote(
  tokens: readonly string[],
  candidates: readonly EngineLanguage[],
): EngineLanguage {
  const hits = new Map<EngineLanguage, number>();
  for (const t of tokens) {
    const langs = STOPWORD_LANGUAGES.get(t);
    if (langs === undefined) continue;
    for (const lang of langs) hits.set(lang, (hits.get(lang) ?? 0) + 1);
  }
  let best = candidates[0]!;
  let bestHits = 0;
  for (const lang of candidates) {
    const h = hits.get(lang) ?? 0;
    if (h > bestHits) {
      bestHits = h;
      best = lang;
    }
  }
  return best;
}

/**
 * Language of a normalised text. The script of its letters decides first:
 * any kana → ja, any Hangul → ko, any Han → zh, any Arabic letter → ar.
 * Cyrillic and Latin texts vote with the stopword lists of the languages of
 * that script (most hits wins, ties in languages.json order, the script's
 * first language when nothing matches: ru and en). Texts whose letters are
 * mostly of another script, or that have no letters, are 'other'.
 */
export function detectLanguage(normalized: string): DetectedLanguage {
  const counts: Record<ScriptBucket, number> = {
    kana: 0,
    hangul: 0,
    han: 0,
    arabic: 0,
    cyrillic: 0,
    latin: 0,
    other: 0,
  };
  for (const ch of normalized) {
    const bucket = scriptOf(ch.codePointAt(0)!);
    if (bucket !== null) counts[bucket] += 1;
  }
  if (counts.kana > 0) return 'ja';
  if (counts.hangul > 0) return 'ko';
  if (counts.han > 0) return 'zh';
  if (counts.arabic > 0) return 'ar';
  if (counts.cyrillic > 0) return stopwordVote(tokenize(normalized), CYRILLIC_LANGUAGES);
  if (counts.latin > 0 && counts.latin >= counts.other) {
    return stopwordVote(tokenize(normalized), LATIN_LANGUAGES);
  }
  return 'other';
}

/** Character n-gram set over code points (spaces included). */
export function charNgrams(normalized: string, n: number): Set<string> {
  const chars = Array.from(normalized);
  const grams = new Set<string>();
  for (let i = 0; i + n <= chars.length; i += 1) {
    grams.add(chars.slice(i, i + n).join(''));
  }
  return grams;
}

export function jaccard(a: Set<string>, b: Set<string>): number {
  if (a.size === 0 && b.size === 0) return 0;
  let inter = 0;
  const [small, large] = a.size <= b.size ? [a, b] : [b, a];
  for (const g of small) if (large.has(g)) inter += 1;
  const union = a.size + b.size - inter;
  return union === 0 ? 0 : inter / union;
}

/** Whole-word substring test on normalised text. */
export function containsPhrase(normalized: string, phrase: string): boolean {
  return ` ${normalized} `.includes(` ${phrase} `);
}

/** Phrase test on normalised text: whole words in 'word' mode, plain substring in 'substring' mode. */
export function matchesPhrase(normalized: string, phrase: string, mode: MatchMode): boolean {
  return mode === 'word' ? containsPhrase(normalized, phrase) : normalized.includes(phrase);
}

File: apps/api/signalyze_api/engine/text.py
"""Text helpers. Line-for-line twin of ``packages/signals/src/text.ts``.

All lengths are counted in Unicode code points so both runtimes agree, except where the
reference deliberately relies on JavaScript's UTF-16 ``String.length`` (see
:func:`utf16_length`).
"""

from __future__ import annotations

import re
import unicodedata
from collections.abc import Sequence
from typing import Literal

from signalyze_api.engine.data_files import STOPWORDS

EngineLanguage = Literal["en", "tr", "de", "es"]
ENGINE_LANGUAGES: tuple[EngineLanguage, ...] = ("en", "tr", "de", "es")

_STOPWORDS: dict[EngineLanguage, frozenset[str]] = {
    "en": frozenset(STOPWORDS.en),
    "tr": frozenset(STOPWORDS.tr),
    "de": frozenset(STOPWORDS.de),
    "es": frozenset(STOPWORDS.es),
}

# Characters removed by JavaScript's String.prototype.trim (WhiteSpace and
# LineTerminator), listed as code points so the source stays ASCII.
_JS_WHITESPACE_CODE_POINTS = (
    0x09,
    0x0A,
    0x0B,
    0x0C,
    0x0D,
    0x20,
    0xA0,
    0x1680,
    0x2000,
    0x2001,
    0x2002,
    0x2003,
    0x2004,
    0x2005,
    0x2006,
    0x2007,
    0x2008,
    0x2009,
    0x200A,
    0x2028,
    0x2029,
    0x202F,
    0x205F,
    0x3000,
    0xFEFF,
)
JS_WHITESPACE = "".join(chr(cp) for cp in _JS_WHITESPACE_CODE_POINTS)

# Python's \w is letters + numbers + underscore, so the underscore is listed explicitly to
# match the reference /[^\p{L}\p{N}\s]+/u. Runs collapse to one space in the next step.
_NON_WORD = re.compile(r"[^\w\s]|_")
_SPACES = re.compile(r"\s+")


def js_trim(text: str) -> str:
    """``String.prototype.trim`` equivalent."""
    return text.strip(JS_WHITESPACE)


def normalize_text(text: str) -> str:
    """NFKC, lower-case, keep letters/digits/whitespace, collapse whitespace, trim."""
    lowered = unicodedata.normalize("NFKC", text).lower()
    spaced = _NON_WORD.sub(" ", lowered)
    return _SPACES.sub(" ", spaced).strip()


def code_point_length(text: str) -> int:
    return len(text)


def utf16_length(text: str) -> int:
    """JavaScript ``String.length`` (UTF-16 code units) for the one place the reference uses it."""
    extra = 0
    for ch in text:
        if ord(ch) > 0xFFFF:
            extra += 1
    return len(text) + extra


def tokenize(normalized: str) -> list[str]:
    if normalized == "":
        return []
    return normalized.split(" ")


def detect_language(tokens: Sequence[str]) -> EngineLanguage:
    """Language whose stopwords occur most often; ties resolve in ENGINE_LANGUAGES order."""
    best: EngineLanguage = "en"
    best_hits = 0
    for lang in ENGINE_LANGUAGES:
        stop = _STOPWORDS[lang]
        hits = 0
        for t in tokens:
            if t in stop:
                hits += 1
        if hits > best_hits:
            best_hits = hits
            best = lang
    return best


def char_ngrams(normalized: str, n: int) -> set[str]:
    """Character n-gram set over code points (spaces included)."""
    return {normalized[i : i + n] for i in range(len(normalized) - n + 1)}


def jaccard(a: set[str], b: set[str]) -> float:
    if len(a) == 0 and len(b) == 0:
        return 0.0
    inter = len(a & b)
    union = len(a) + len(b) - inter
    return 0.0 if union == 0 else inter / union


def contains_phrase(normalized: str, phrase: str) -> bool:
    """Whole-word substring test on normalised text."""
    return f" {phrase} " in f" {normalized} "
